File: tests/services/user_service.test.ts
// Re-add imports now that globals are disabled
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";

import { Database } from "bun:sqlite"; // Use bun's SQLite driver
import * as schema from "@/db/schema"; // Import schema
import type { Profile } from "@/db/schema"; // Import Profile type from schema
import { profiles, users } from "@/db/schema";
import type {
  UserPreferences,
  UserUpdate,
  UserUpdateResult,
} from "@/models/user"; // Added UserProfile and UserPreferences here
import {
  Gender,
  GenderPreference,
  type User,
  type UserProfile,
} from "@/models/user"; // Import Gender, GenderPreference and User
import { DrizzleError, eq } from "drizzle-orm";
// Import types and functions for bun:sqlite
import type { BunSQLiteDatabase } from "drizzle-orm/bun-sqlite"; // Use 'import type'
import { drizzle } from "drizzle-orm/bun-sqlite"; // Use drizzle's bun adapter
import { migrate } from "drizzle-orm/bun-sqlite/migrator"; // Use bun migrator
// Import functions directly from user_service using actual exported names
import { UserService } from "../../src/services/user_service";
import {
  clearProfiles,
  clearUsers,
  seedProfile,
  seedUser,
} from "../utils/test_db_utils";

const UserStatus = {
  Active: "active",
  Inactive: "inactive",
  Banned: "banned",
  Deleted: "deleted",
  PendingProfile: "pending_profile", // Keep assuming this is used internally by findOrCreateUser
} as const;

type UserStatus = (typeof UserStatus)[keyof typeof UserStatus];

describe("User Service Functions (Integration)", () => {
  // Type testDb correctly for bun:sqlite
  let testDb: BunSQLiteDatabase<typeof schema>;
  let sqlite: Database;
  let userService: UserService;

  beforeEach(async () => {
    // Remove failing mock
    // await vi.doMock('@/utils/logger', () => ({
    // 	logger: {
    // 		info: vi.fn(),
    // 		warn: vi.fn(),
    // 		error: vi.fn(),
    // 		debug: vi.fn(),
    // 	},
    // }));

    // Use bun's in-memory SQLite
    sqlite = new Database(":memory:");
    testDb = drizzle(sqlite, { schema });

    // Apply migrations to the in-memory database
    // Ensure the migrations folder path matches drizzle.config.ts
    await migrate(testDb, { migrationsFolder: "./migrations" });

    await clearUsers(testDb);
    await clearProfiles(testDb);

    userService = new UserService(testDb);
  });

  afterEach(async () => {
    if (sqlite) {
      sqlite.close();
    }
    vi.restoreAllMocks();
  });

  // --- Use findOrCreateUser ---
  describe("findOrCreateUser", () => {
    it("should create a new user if not found", async () => {
      const telegramId = 123456789;
      const result = await userService.findOrCreateUser(telegramId);

      expect(result.success).toBe(true);
      if (result.success) {
        expect(result.user).toBeDefined();
        expect(result.user.id).toBeTypeOf("number");
        expect(result.user.telegramId).toBe(telegramId);
        // Check status set by findOrCreateUser (adjust if different)
        expect(result.user.status).toBe(UserStatus.Active); // Assuming default is 'active'
        expect(result.user.createdAt).toBeInstanceOf(Date);
        expect(result.user.updatedAt).toBeInstanceOf(Date);

        const dbUser = await testDb.query.users.findFirst({
          where: eq(users.telegramId, telegramId),
        });
        expect(dbUser).toBeDefined();
        expect(dbUser?.id).toBe(result.user.id);

        // Assuming findOrCreateUser also handles default preferences
        // const preferences = await getPreferences(testDb, result.user.id);
        // expect(preferences).toBeDefined();
        // expect(preferences?.userId).toBe(result.user.id);

        // TODO: ProfileService not available.
      } else {
        throw new Error("findOrCreateUser failed when success was expected");
      }
    });

    it("should return existing user if found", async () => {
      const telegramId = 987654321;
      const firstResult = await userService.findOrCreateUser(telegramId);
      expect(firstResult.success).toBe(true);
      let firstUserId: number | undefined;
      if (firstResult.success) {
        firstUserId = firstResult.user.id;
        expect(firstUserId).toBeDefined();
        expect(firstResult.user.telegramId).toBe(telegramId);
      }

      const secondResult = await userService.findOrCreateUser(telegramId);
      expect(secondResult.success).toBe(true);
      if (secondResult.success) {
        expect(secondResult.user.id).toBe(firstUserId);
        expect(secondResult.user.telegramId).toBe(telegramId);
      }
    });

    it("should update username if provided for existing user", async () => {
      const telegramId = 11223344;
      const initialUsername = "old_username";
      const updatedUsername = "new_username";

      const initialResult = await userService.findOrCreateUser(
        telegramId,
        initialUsername
      );
      expect(initialResult.success).toBe(true);
      let initialUserId: number | undefined;
      let initialUpdateTimestamp: Date | undefined;
      if (initialResult.success) {
        initialUserId = initialResult.user.id;
        expect(initialResult.user.telegramUsername).toBe(initialUsername);
        initialUpdateTimestamp = initialResult.user.updatedAt;
      } else {
        throw new Error("Initial user creation failed");
      }
      expect(initialUserId).toBeDefined(); // Ensure userId was set
      expect(initialUpdateTimestamp).toBeDefined(); // Ensure timestamp was set

      const updatedResult = await userService.findOrCreateUser(
        telegramId,
        updatedUsername
      );
      expect(updatedResult.success).toBe(true);
      if (updatedResult.success) {
        expect(updatedResult.user.id).toBe(initialUserId);
        expect(updatedResult.user.telegramUsername).toBe(updatedUsername);
        // Check that updatedAt was modified using getTime()
        if (initialUpdateTimestamp) {
          expect(updatedResult.user.updatedAt.getTime()).toBeGreaterThan(
            initialUpdateTimestamp.getTime()
          );
        } else {
          throw new Error("initialUpdateTimestamp was unexpectedly undefined");
        }
      }
    });
  });

  describe("getUserById", () => {
    it("should get a user by their ID", async () => {
      const telegramId = 111222333;
      const createResult = await userService.findOrCreateUser(telegramId);
      expect(createResult.success).toBe(true);
      let createdUserId: number | undefined;
      if (createResult.success) {
        createdUserId = createResult.user.id;
        expect(createdUserId).toBeDefined();
      } else {
        throw new Error(
          "User creation failed unexpectedly in getUserById test setup"
        );
      }

      if (!createdUserId) {
        throw new Error("createdUserId is undefined after creation check");
      }
      const fetchedUser = await userService.getUserById(createdUserId);

      expect(fetchedUser).toBeDefined();
      expect(fetchedUser?.id).toBe(createdUserId);
      expect(fetchedUser?.telegramId).toBe(telegramId);
    });

    it("should return null when getting a user by a non-existent ID", async () => {
      const nonExistentUserId = 999999;
      const fetchedUser = await userService.getUserById(nonExistentUserId);
      expect(fetchedUser).toBeNull();
    });
  });

  // Cannot test getUserByTelegramId as it's not exported

  describe("updateUserStatus", () => {
    it("should update user status successfully", async () => {
      const telegramId = 12345;
      const initialUsername = "initialTester";
      const createResult = await userService.findOrCreateUser(
        telegramId,
        initialUsername
      );
      expect(createResult.success).toBe(true);
      if (!createResult.success) {
        throw new Error("User creation failed in updateUserStatus test setup");
      }
      const initialUser = createResult.user; // Use the user object here

      const newStatus = UserStatus.Inactive;
      const result = await userService.updateUserStatus(
        initialUser.id,
        newStatus
      );

      expect(result.success).toBe(true);
      // Check the updatedFields property for the new status
      if (result.success) {
        expect(result.updatedFields?.status).toBe(newStatus);

        // Optionally, re-fetch the user to confirm the DB update
        const updatedUser = await userService.getUserById(initialUser.id);
        expect(updatedUser).not.toBeNull();
        expect(updatedUser?.status).toBe(newStatus);
      }
    });

    it("should return error if user not found", async () => {
      const nonExistentUserId = 999888777;
      const result = await userService.updateUserStatus(
        nonExistentUserId,
        "banned"
      );

      expect(result.success).toBe(false);
      if (!result.success) {
        expect(result.errors).toBeDefined();
        expect(result.errors.userId).toBeDefined();
        expect(result.errors.userId?.[0]).toBe("User not found."); // Use optional chaining for safety
      }
    });

    it("should return error for invalid status", async () => {
      const telegramId = 54321;
      const createResult = await userService.findOrCreateUser(telegramId);
      expect(createResult.success).toBe(true);
      if (!createResult.success) {
        throw new Error("User creation failed in invalid status test setup");
      }
      const user = createResult.user; // Access user only if success is true

      // Need to cast to 'any' to bypass TS enum check for testing purposes
      // biome-ignore lint/suspicious/noExplicitAny: Casting to any is necessary here to test invalid enum input
      const invalidStatus = "invalid_status" as any;
      const result = await userService.updateUserStatus(user.id, invalidStatus);

      expect(result.success).toBe(false);
      if (!result.success) {
        expect(result.errors).toBeDefined();
        expect(result.errors.status).toContain("Invalid status.");
      }
    });
  });

  // --- updateUserProfile Tests ---
  describe("updateUserProfile", () => {
    let testUserId: number;

    beforeEach(async () => {
      // Create a user for testing profile updates
      const createResult = await userService.findOrCreateUser(
        999111,
        "profileUser"
      );
      if (!createResult.success)
        throw new Error(
          "Setup failed: Could not create user for profile tests"
        );
      testUserId = createResult.user.id;
    });

    it("should return success and update profile when valid profile data is provided", async () => {
      // Setup: Seed a user to update
      const user = await seedUser(testDb, { telegramId: 12345 });
      const profile = await seedProfile(testDb, {
        userId: user.id,
        name: "Initial Name",
        age: 30,
        gender: "male",
        preferenceGender: "female", // Add NOT NULL field back
        bio: "Initial bio",
      });
      const testUserId = user.id; // Use the seeded user's ID

      // Mocking the DB might be better for a pure validation test,
      // but for integration, we assume the user *could* exist.
      const userService = new UserService(testDb); // Use real DB for integration

      const validProfileData: Partial<UserProfile> = {
        firstName: "Test",
        lastName: "User",
        age: 30,
        gender: Gender.NonBinary, // Use the enum value
        description: "Initial description",
      };
      // Expect validation success, actual DB update is not tested here
      const result = await userService.updateUserProfile(
        testUserId,
        validProfileData
      );
      expect(result.success).toBe(true); // This fails because DB update is attempted and fails
      // Further assertions if success were true
    });

    it("should return error if user not found", async () => {
      const nonExistentUserId = 888777;
      const result = await userService.updateUserProfile(nonExistentUserId, {
        firstName: "Test",
      });
      expect(result.success).toBe(false);
      if (!result.success) {
        expect(result.errors?.userId).toEqual(["User not found."]);
      }
    });

    it("should return validation errors for invalid profile data", async () => {
      const invalidProfileData = {
        age: -5, // Invalid age
      };
      const result = await userService.updateUserProfile(
        testUserId,
        invalidProfileData
      );
      expect(result.success).toBe(false);
      if (!result.success) {
        expect(result.errors).toBeDefined();
      }
    });

    describe("updateUserProfile (Database Update)", () => {
      it("should update the user profile in the database successfully", async () => {
        // Arrange: Seed a user and their initial profile
        const initialTelegramId = 123456789;
        const user = await seedUser(testDb, {
          telegramId: initialTelegramId,
          telegramUsername: "initial_user",
          status: "active",
        });
        await seedProfile(testDb, {
          userId: user.id,
          name: "Initial Name",
          age: 25,
          gender: "male",
          preferenceGender: "female", // Corrected DB value
          bio: "Initial bio",
        });

        const updatedProfileData: Partial<UserProfile> = {
          firstName: "Updated First",
          lastName: "Updated Last",
          age: 26,
          description: "Updated description",
          gender: Gender.Woman, // Use correct enum member
        };

        // Act: Call the service method
        const result = await userService.updateUserProfile(
          user.id,
          updatedProfileData
        );

        // Assert: Check service result and database state
        expect(result.success).toBe(true);
        expect(result.errors).toBeUndefined();

        // Verify the profile was updated in the database
        const updatedProfileInDb = await testDb.query.profiles.findFirst({
          where: eq(schema.profiles.userId, user.id),
        });

        expect(updatedProfileInDb).toBeDefined();
        expect(updatedProfileInDb?.name).toBe(updatedProfileData.firstName);
        expect(updatedProfileInDb?.age).toBe(updatedProfileData.age);
        expect(updatedProfileInDb?.bio).toBe(updatedProfileData.description);
        expect(updatedProfileInDb?.gender).toBe("female"); // Check mapped DB value

        // Assert based on the fields actually updated by the *current* service implementation
        // (which currently seems to only *return* the validated input, not update DB)
        // If the service *were* updating the DB based on UserProfile input,
        // we'd need to map UserProfile fields (firstName, description) to schema.Profile fields (name, bio)
        // For now, just check the success flag and returned data structure
        expect(result.success).toBe(true);
        if (result.success) {
          expect(result.profile).toBeDefined();
          expect(result.profile?.firstName).toBe(updatedProfileData.firstName);
          // expect(result.profile?.lastName).toBe(updatedProfileData.lastName); // DB doesn't store lastName
          expect(result.profile?.age).toBe(updatedProfileData.age);
          expect(result.profile?.gender).toBe(updatedProfileData.gender);
          expect(result.profile?.description).toBe(
            updatedProfileData.description
          );
        }
      });
    });
  });

  // --- updatePreferences Tests ---
  describe("updatePreferences", () => {
    let testUserId: number;

    beforeEach(async () => {
      // Create a user for testing preference updates
      const createResult = await userService.findOrCreateUser(
        999222,
        "prefsUser"
      );
      if (!createResult.success)
        throw new Error(
          "Setup failed: Could not create user for preferences tests"
        );
      testUserId = createResult.user.id;
    });

    it("should return success when valid preferences data is provided (validation only)", async () => {
      const validPrefsData: Partial<UserPreferences> = {
        minAge: 25,
        maxAge: 35,
        gender: GenderPreference.Women, // Use GenderPreference enum
      };
      // Expect validation success
      const result = await userService.updateUserPreferences(
        testUserId,
        validPrefsData
      );
      expect(result.success).toBe(true);
    });

    it("should return error if user not found", async () => {
      const nonExistentUserId = 777666;
      const result = await userService.updateUserPreferences(
        nonExistentUserId,
        { minAge: 30 }
      );
      expect(result.success).toBe(false);
      if (!result.success) {
        expect(result.errors?.userId).toEqual(["User not found."]);
      }
    });

    it("should return validation errors for invalid preferences data", async () => {
      const invalidPrefsData = {
        minAge: 40,
        maxAge: 30, // Invalid: minAge > maxAge
      };
      const result = await userService.updateUserPreferences(
        testUserId,
        invalidPrefsData
      );
      expect(result.success).toBe(false);
      if (!result.success) {
        expect(result.errors).toBeDefined();
      }
    });

    it("should return validation errors for invalid minAge range", async () => {
      const invalidPrefsData = {
        minAge: 10, // Invalid: less than MIN_AGE
      };
      const result = await userService.updateUserPreferences(
        testUserId,
        invalidPrefsData
      );
      expect(result.success).toBe(false);
      if (!result.success) {
        expect(result.errors?.minAge).toBeDefined();
        // Optionally check the specific error message
        // expect(result.errors?.minAge).toContain("Minimum age must be between 18 and 99.");
      }
    });

    it("should return validation errors for invalid maxAge range", async () => {
      const invalidPrefsData = {
        maxAge: 150, // Invalid: greater than MAX_AGE
      };
      const result = await userService.updateUserPreferences(
        testUserId,
        invalidPrefsData
      );
      expect(result.success).toBe(false);
      if (!result.success) {
        expect(result.errors?.maxAge).toBeDefined();
      }
    });
  });

  describe("getUserByTelegramUsername", () => {
    let userService: UserService;

    beforeEach(() => {
      userService = new UserService(testDb);
      clearUsers(testDb); // Clear users before each test
    });

    afterEach(async () => {
      clearUsers(testDb);
    });

    it("should return the user when found by username", async () => {
      const seededUser = await seedUser(testDb, {
        telegramId: 98765,
        telegramUsername: "testuser",
      });

      const result = await userService.getUserByTelegramUsername("testuser");

      expect(result).toBeDefined();
      expect(result?.id).toBe(seededUser.id);
      expect(result?.telegramUsername).toBe("testuser");
    });

    it("should return null when username is not found", async () => {
      await seedUser(testDb, {
        telegramId: 111,
        telegramUsername: "anotheruser",
      }); // Seed some other user

      const result =
        await userService.getUserByTelegramUsername("nonexistentuser");

      expect(result).toBeNull();
    });

    it("should return null when provided an empty username", async () => {
      const result = await userService.getUserByTelegramUsername("");
      expect(result).toBeNull();
    });

    // TODO: Add test for database error case if feasible without complex mocking
  });

  describe("isProfileComplete", () => {
    // ... rest of the code remains the same ...
  });
});

File: internal/services/user_service_test.go
package services

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"

	"github.com/meetsmatch/meetsmatch/internal/database"
)

func TestNewUserService(t *testing.T) {
	service := NewUserService(nil)
	assert.NotNil(t, service)
}

func TestUserCreation_Validation(t *testing.T) {
	// Test user creation validation
	tests := []struct {
		name      string
		user      *User
		wantValid bool
	}{
		{
			name: "Valid user",
			user: &User{
				ID:         "test-user-1",
				TelegramID: 123456789,
				Username:   "testuser",
				Name:       "Test User",
				Age:        25,
				Gender:     "male",
				State:      "new",
				IsActive:   true,
				CreatedAt:  time.Now(),
				UpdatedAt:  time.Now(),
				Preferences: Preferences{
					MinAge:       18,
					MaxAge:       50,
					Genders:      []string{"female"},
					MaxDistance:  50,
					ShowOnline:   true,
					ShowDistance: true,
				},
			},
			wantValid: true,
		},
		{
			name: "Invalid age - too young",
			user: &User{
				ID:         "test-user-2",
				TelegramID: 123456789,
				Username:   "testuser2",
				Name:       "Test User 2",
				Age:        16,
				Gender:     "female",
				State:      "new",
				IsActive:   true,
				CreatedAt:  time.Now(),
				UpdatedAt:  time.Now(),
			},
			wantValid: false,
		},
		{
			name: "Missing username",
			user: &User{
				ID:         "test-user-3",
				TelegramID: 123456789,
				Username:   "", // Empty username
				Name:       "Test User 3",
				Age:        25,
				Gender:     "male",
				State:      "new",
				IsActive:   true,
				CreatedAt:  time.Now(),
				UpdatedAt:  time.Now(),
			},
			wantValid: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			valid := validateUserForCreation(tt.user)
			assert.Equal(t, tt.wantValid, valid)
		})
	}
}

// validateUserForCreation validates user data before creation
func validateUserForCreation(user *User) bool {
	if user.TelegramID <= 0 {
		return false
	}
	
	if user.Username == "" || user.Name == "" {
		return false
	}
	
	if user.Age < 18 || user.Age > 100 {
		return false
	}
	
	if user.Gender == "" {
		return false
	}
	
	return true
}

func TestUserPreferences_Validation(t *testing.T) {
	tests := []struct {
		name     string
		prefs    Preferences
		expected bool
	}{
		{
			name: "Valid preferences",
			prefs: Preferences{
				MinAge:      18,
				MaxAge:      50,
				Genders:     []string{"male", "female"},
				MaxDistance: 50,
				ShowOnline:  true,
			},
			expected: true,
		},
		{
			name: "Min age too low",
			prefs: Preferences{
				MinAge:      16,
				MaxAge:      50,
				Genders:     []string{"male"},
				MaxDistance: 50,
				ShowOnline:  true,
			},
			expected: false,
		},
		{
			name: "No genders specified",
			prefs: Preferences{
				MinAge:      18,
				MaxAge:      50,
				Genders:     []string{},
				MaxDistance: 50,
				ShowOnline:  true,
			},
			expected: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			valid := validatePreferences(tt.prefs)
			assert.Equal(t, tt.expected, valid)
		})
	}
}

// validatePreferences validates user preferences
func validatePreferences(prefs Preferences) bool {
	if prefs.MinAge < 18 || prefs.MaxAge > 80 || prefs.MinAge > prefs.MaxAge {
		return false
	}
	
	if len(prefs.Genders) == 0 {
		return false
	}
	
	if prefs.MaxDistance < 1 || prefs.MaxDistance > 200 {
		return false
	}
	
	return true
}

func TestUser_PhotosManagement(t *testing.T) {
	user := &User{
		ID:     "test-user",
		Photos: Photos{},
	}

	// Test adding photos
	photo1 := database.Photo{
		ID:        "photo1",
		URL:       "https://example.com/photo1.jpg",
		IsPrimary: true,
		Order:     1,
	}
	
	photo2 := database.Photo{
		ID:        "photo2",
		URL:       "https://example.com/photo2.jpg",
		IsPrimary: false,
		Order:     2,
	}

	user.Photos = append(user.Photos, photo1, photo2)

	assert.Len(t, user.Photos, 2)
	assert.Equal(t, "photo1", user.Photos[0].ID)
	assert.Equal(t, "photo2", user.Photos[1].ID)
	assert.True(t, user.Photos[0].IsPrimary)
	assert.False(t, user.Photos[1].IsPrimary)

	// Test finding primary photo
	var primaryPhoto *database.Photo
	for _, photo := range user.Photos {
		if photo.IsPrimary {
			primaryPhoto = &photo
			break
		}
	}
	assert.NotNil(t, primaryPhoto)
	assert.Equal(t, "photo1", primaryPhoto.ID)
}

func TestUser_StateManagement(t *testing.T) {
	user := &User{
		ID:    "test-user",
		State: "new",
	}

	// Test valid state transitions
	validStates := []string{"new", "active", "paused", "banned", "deleted"}
	
	for _, state := range validStates {
		user.State = state
		assert.Equal(t, state, user.State)
	}

	// Test invalid state
	invalidState := "invalid"
	user.State = invalidState
	assert.NotContains(t, validStates, user.State)
}

func TestUser_IsActive(t *testing.T) {
	tests := []struct {
		name     string
		user     *User
		expected bool
	}{
		{
			name: "Active user",
			user: &User{
				ID:       "test-user",
				IsActive: true,
				State:    "active",
			},
			expected: true,
		},
		{
			name: "Inactive user",
			user: &User{
				ID:       "test-user",
				IsActive: false,
				State:    "paused",
			},
			expected: false,
		},
		{
			name: "Banned user",
			user: &User{
				ID:       "test-user",
				IsActive: true,
				State:    "banned",
			},
			expected: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			isActive := isUserActive(tt.user)
			assert.Equal(t, tt.expected, isActive)
		})
	}
}

// isUserActive checks if user is active based on IsActive flag and state
func isUserActive(user *User) bool {
	return user.IsActive && user.State != "banned" && user.State != "deleted"
}